### ISSUE TITLE
pam_u2f and first login (after reboot/shutdown) - no login, no fail

### ISSUE THREAD

**Issue by elegos** [sddm/sddm#1103]
Hello! I'm running Fedora 29 with KDE and sddm. I've got a Yubico SecurityKey (hardware key) which handles both U2F and FIDO 2.0 protocols. I'm using the key as a U2F hardware key to access to my account on the linux box.

I've added the following line to the `/etc/pam.d/system-auth` file:

```
auth        required                                     pam_u2f.so cue authfile=/etc/security/u2f_keys
```
Right before the `pam_unix.so` auth line.

The problem is that at the first login (after shutdown / reboot), `sddm` will prompt me for the password, but pressing `ENTER` won't prompt me to execute the U2F action either, stalling the login (I can still edit the password, as if I never pressed `ENTER`). Logging in via TTY works correctly (it asks for U2F and then for password). I have to set the auth to `sufficient` (or comment the line) and reboot, disabling U2F, in order to enter my KDE session again.

After the 1st login I can restore the line (auth required) and lock the session or terminate it, then I can correctly enter the user environment because it will prompt me for the U2F action.

**Steps to reproduce**

0. (setup a U2F profile for the user)
1. enable pam_u2f.so in `/etc/pam.d/system-auth` as required and with `cue` option before the pam_unix auth entry
2. restart
3. try to login

**Comment by AshenColors** [comment]
Gonna blow the dust off this one and report that I ran into this with Fedora Kinoite 35, a new immutable distribution based on Silverblue that features KDE. SDDM does not function correctly when running into Authselect's `pam_u2f.so cue` and won't accept input from a hardware authenticator that otherwise works. The description of the bug above is still accurate.

Plasma's lockscreen handles `pam_u2f.so cue` just fine, displaying the cue text and everything, so it seems like it's definitely intended functionality within the KDE desktop environment. Silverblue/Kinoite, being an ostree project, can easily rebase itself on an image with a different desktop environment, which makes it even more likely that a user with existing 2FA workflows configured will run into SDDM and find themselves staring at a frozen display instead. 

Right now, the only known workaround appears to be to swap in GDM or another display manager of choice.

edit: since it's not immediately obvious, pam-u2f actually does handle FIDO2 and not (just) the now-deprecated-by-Chrome U2F. They retained the name for legacy compatibility.

**Comment by firelizzard18** [comment]
I'm running into the same issue with Gentoo. `auth sufficient pam_u2f.so` (without `cue`) allows me to login, but it **requires** U2F in addition to my password when it should be **instead of** my password. And if I add `cue` it just breaks and I can't login (I have to use a TTY to update the configuration and `systemctl restart sddm`).

**Comment by ashleysommer** [comment]
I'm running into this issue too, on Manjaro-Plasma. When cue is used, SDDM does not progress the login once the button on the device is pressed.

@firelizzard18 I believe you can simply leave the password blank, press enter, then tap the device. You don't need to put our actual password in.

**Comment by nagmat84** [comment]
I would like to blow of the dust of this issue as I encountered a probably related problem today (see [Gentoo Forums: Desktop Environments -> SDDM with FIDO2 authentication: Cue is missing](https://forums.gentoo.org/viewtopic-p-8827078.html#8827078)).

My PAM configuration for `system-auth` is:
```
auth   required                                                   pam_env.so
auth   requisite                                                  pam_faillock.so   preauth
auth   [success=2 new_authtok_reqd=2 default=ignore]              pam_u2f.so        cue
auth   [success=1 new_authtok_reqd=1 ignore=ignore default=bad]   pam_unix.so       nullok try_first_pass
auth   [default=die]                                              pam_faillock.so   authfail
auth   optional                                                   pam_cap.so
```
The idea is that `pam_u2f` takes precedence over traditional password-based login, but pasword-based login should still be available as a fallback option:

 * If `pam_u2f` succeeds, PAM skip the next 2 lines, executes `pam_cap` and returns `success`.
 * If `pam_u2f` fails, PAM ignores the failure and continues with `pam_unix` as usual.

This works for console-based login as expected.

With SDDM I need to touch the FIDO2 stick _twice_ for a successful login. More precisely, I see the following odd behavior:

 1. SDDM prompts me for a password; SDDM does _not_ print the cue and the LED of the FIDO2 stick does not flash
 2. When I touch the FIDO2 stick for the _1st_ time, characters are typed into the password input field of SDDM. I cannot tell what characters, because they are replaced by asterisks, but it seems as if the FIDO2 stick acts like a keyboard
 3. The (password-based) login fails due to the (presumably) wrong password.  The password input field remains deactivated, but the LED of the FIDO2 stick starts flashing.
 4. When I touch the FIDO2 stick for the _2nd_ time, the login succeeds.

I have no idea what is going on under the hood, but it seems to be the following: When SDDM calls PAM for authentication the first time, `pam_u2f` somehow fails. Hence, SDDM falls back to password-based login. This is in line with the other reports in this bug reports except that my PAM configuration allows the fallback option. As the FIDO2 stick is not properly initialized (the LED is not flashing), it behaves like a keyboard. When the 1st password-based authentication attempt fails (expect that the FIDO2 stick enters the "password"), SDDM restartst the PAM authentication procedure. This 2nd time FIDO2 authentication works and I can login.

**Comment by Vogtinator** [comment]
SDDM does not run any PAM code until you click the login button. You can just hit enter without typing a password and pam_u2f will run as usual.

**Comment by Vogtinator** [comment]
IMO this is basically a duplicate of #284.

**Comment by nagmat84** [comment]
Indeed. However, it would be nice if SDDM would not require to press the login button first, before it starts the PAM authentication. It would also be nice to see the cue (i.e. "touch security device") in the first place.

I don't know how, but the KDE lock screen and other display managers (like GDM) are able to do that.

**Comment by Vogtinator** [comment]
See #284 and the discussion on #1220.
